### PR TITLE
[server] Introduce RequestContext

### PR DESCRIPTION
**Pull request by geropl**
## Description
Context:
 - Project: https://linear.app/gitpod/project/fga-compatible-api-tokens-a27f6bdb53d2
 - Project Spec: https://www.notion.so/gitpod/FGA-compatible-API-Tokens-Project-Spec-7d47f66fab4c471da6429a90dfbd45e7

Introduces `RequestContext` as primary method to pass request-local information between different pieces of code, incl. for authorization and request-local caching. This PR makes sure all code that calls some `...Service` "request-handling" logic in the end is wrapped inside a `runWithRequestContext` call:
 - websocket connections
 - HTTP requests to various express apps
 - gRPC services
This also means that whenever we execute code with other permissions, we need to call `runWithSubjectId` (in the future). See the call sites in this PR for examples.

To access the request based context, the unit exports a number of functions prefixed by `ctx...`, e.g.:
 - `ctxUserId`
 - `ctxCheckAborted`
The idea for those is to be used _sparingly_, and almost exclusively in "leaf" code (authorizer, logging, tracing, etc.).

To pass authentication information independent of the `User`/`userId` string, a new `SubjectId` type is introduced. Currently there is only one `SubjectIdKind` ("user"), which serializes to: `user_<id>`.
This is expected to be extended by a kind `apitokenv0_<id>` in subsequent PRs.

### Follow-up work
- add metrics to measure that we added `RequestContexts` everywhere
- switch authorization from "first argument (userId)" to `ctx.subjectId`. Not passing that will be an "access denied" error.
- In consequence this means that in future PRs, we are going to remove the first argument `userId: string` that we are currently passing into `...Service`s and down to the `Authorizer`. For now, in order to allow to split up work, and not interfere too much with other API-related work, we still pass that argument in the new top-level code (gRPC servicve implementations) using `ctxUserId`.

I will send out additional information about the future steps summarized here, and the expected changes to our code (patterns) today.

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at c3bde8b</samp>

This pull request introduces a new way of identifying users and organizations in the analytics system using subject ids, which are more reliable and consistent than user ids. It also refactors the request context mechanism to use subject ids and provide more information about the requests and the users. It updates various components and files to use the new request context and subject id features.

</details>

## Related Issue(s)

Fixes EXP-915

## How to test

For this PR, we need to make sure we don't break existing things.
Most interesting for review:
 - Authorizer
 - SpiceDB authorizer
 - run all tests (CI)
 - all changes the new gRPC code!

For testing:
 - check that logging still works as expected (and carries the [additional metadata](https://github.com/gitpod-io/gitpod/pull/19023/files#diff-1cf08c165f36290d5e6e2f5373a3f398e9f55d41c4a888cdfb9d8c0de19c6ec5R15))
 - verify the changes to 

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - gpl-request-context</li>
	<li><b>🔗 URL</b> - <a href="https://gpl-request-context.preview.gitpod-dev.com/workspaces" target="_blank">gpl-request-context.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - gpl-request-context-gha.19488</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-gpl-request-context%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [x] analytics=log
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/api/server.ts
Comment (geropl) [inline components/server/src/api/server.ts:233]: @akosyakov This is the biggest change here: before we would call `verify` (authentication + ratelimit) on every generation. I'm pretty sure we don't want authentication in there. But maybe ratelimiting? WDYT? :shrug:

File: components/server/src/api/server.ts
Comment (akosyakov) [inline components/server/src/api/server.ts:211]: does server streaming support `Promise<AsyncGenerator>`, I thought proxy should return AsyncGenerator?

Comment (geropl) [reply components/server/src/api/server.ts:211]: Uh, you are right: https://github.com/connectrpc/connect-es/blob/main/packages/connect/src/implementation.ts#L212C12-L212C12
Also, just makes sense... let me see how I can make that work...

File: components/server/src/api/hello-service-api.ts
Comment (akosyakov) [inline components/server/src/api/hello-service-api.ts:20]: We can completely remove the second argument everywhere?

File: components/server/src/api/server.ts
Comment (akosyakov) [reply components/server/src/api/server.ts:233]: it is addressed now, right? Or misunderstand something.

I think we call auth only once before creating a generator

**Review by akosyakov — COMMENTED** [review]
I skimmed through the code it looks reasonable. cc @svenefftinge to have a look as well

**Comment by geropl** [comment]
> it is addressed now, right? Or misunderstand something.
>
> I think we call auth only once before creating a generator

Yes, I pushed an attempt to fix it, by making `auth` an explicit, separate step. :+1:

File: components/server/src/api/server.ts
Comment (geropl) [reply components/server/src/api/server.ts:233]: Pushed a fix.

File: components/server/src/api/hello-service-api.ts
Comment (svenefftinge) [inline components/server/src/api/hello-service-api.ts:27]: I would prefer a dedicated function per aspect instead of a generic ctx() that returns everything in one shape.

File: components/server/src/api/organization-service-api.ts
Comment (svenefftinge) [inline components/server/src/api/organization-service-api.ts:56]: same here. I'd prefer a global `currentUserId()` that encapsulates the access to the ctx.

Comment (svenefftinge) [reply components/server/src/api/organization-service-api.ts:56]: I see you have introduced that and marked it as deprecated. I think the function declaration is actually good (incl. the excpetion it throws) but we should not require the first argument on all service methods in the long run (i.e. deprecate that somehow)

File: components/server/src/auth/subject-id.ts
Comment (svenefftinge) [inline components/server/src/auth/subject-id.ts:24]: Would be great to not use a class but only the shape (and call that `SubjectId`). Seems the constructor is not used anywhere else and the serialize and parse could be static.

File: components/server/src/auth/subject-id.ts
Comment (svenefftinge) [inline components/server/src/auth/subject-id.ts:85]: this doesn't seem to be used in production code

File: components/server/src/util/request-context.ts
Comment (svenefftinge) [inline components/server/src/util/request-context.ts:96]: I think we rather want to create a new childContext when we create a span (i.e. the other way around).

File: components/server/src/util/request-context.ts
Comment (svenefftinge) [inline components/server/src/util/request-context.ts:97]: we need to first apply `parent` and then override with child

File: components/server/src/api/server.ts
Comment (geropl) [reply components/server/src/api/server.ts]: Done.

File: components/server/src/auth/subject-id.ts
Comment (geropl) [reply components/server/src/auth/subject-id.ts:24]: We discussed and agreed on the class only, because of ease of usability.

File: components/server/src/auth/subject-id.ts
Comment (geropl) [reply components/server/src/auth/subject-id.ts:85]: True, that is a leftover from an earlier experiment.

**Comment by mustard-mh** [comment]
> AlexTugarev mustard-mh For reading the description, and getting acquainted with the changes to APIs (e.g. not using HandlerContext, but ctxUserId - but only until we drop the need for that). [[ctx](https://github.com/gitpod-io/gitpod/pull/19023#issuecomment-1809741726)]

cc @jeanp413 since you have onboard grpc migration too

**Comment by akosyakov** [comment]
@geropl Could you rebase please? 🙏

**Comment by geropl** [comment]
> @geropl Could you rebase please? 🙏

@akosyakov Done (again :running_man: :wink: ). [Build](https://github.com/gitpod-io/gitpod/actions/runs/6864532750) is still running.

File: components/server/src/util/request-context.ts
Comment (akosyakov) [inline components/server/src/util/request-context.ts:106]: I would name it Cancelled as per gRPC, with Cancelled mesage

Comment (akosyakov) [reply components/server/src/util/request-context.ts:106]: Althoug our internal error codes are http based :face_palm:

Comment (geropl) [reply components/server/src/util/request-context.ts:106]: > Althoug our internal error codes are http based :face_palm:

:100: Followed the existing convention. At least it's a 1-1 translation to "Cancelled". :+1:

File: components/server/src/api/auth-provider-service-api.ts
Comment (akosyakov) [inline components/server/src/api/auth-provider-service-api.ts:45]: there is it used?

Comment (geropl) [reply components/server/src/api/auth-provider-service-api.ts:45]: What do you mean? :thinking: 
I think there is no use-case for accessing HandlerContext in actual service implementations, no?

File: components/server/src/util/request-context.ts
Comment (akosyakov) [inline components/server/src/util/request-context.ts:104]: nice, looking forward to see it integrated into service/DB layer

File: components/server/src/api/organization-service-api.ts
Comment (akosyakov) [inline components/server/src/api/organization-service-api.ts:262]: @gero here 3 argument should be update

File: components/server/src/api/server.ts
Comment (akosyakov) [inline components/server/src/api/server.ts:264]: I'm not 100% sure it will be exected with subjectId, but we will learn it in PRs there we make use of streamin APIs and can fix there.

Comment (akosyakov) [reply components/server/src/api/server.ts:264]: Although I assume you tested with dummy service and results there returned a proepr subhect id

File: components/server/src/prebuilds/github-app.ts
Comment (akosyakov) [inline components/server/src/prebuilds/github-app.ts:128]: do we need to call next here?

File: components/server/src/api/server.ts
Comment (geropl) [reply components/server/src/api/server.ts:264]: I will re-do that. Last time might have been with the previous implementation. :grimacing: :+1:

File: components/server/src/prebuilds/github-app.ts
Comment (geropl) [reply components/server/src/prebuilds/github-app.ts:128]: No, that's a bug! :bug:

File: components/server/src/websocket/websocket-connection-manager.ts
Comment (geropl) [reply components/server/src/websocket/websocket-connection-manager.ts:380]: Ah, thank you! Will adjust.

File: components/server/src/api/auth-provider-service-api.ts
Comment (akosyakov) [reply components/server/src/api/auth-provider-service-api.ts:45]: I mean ownerId is not used in the code below, instead we use ctxUserId

File: components/server/src/util/request-context.ts
Comment (geropl) [reply components/server/src/util/request-context.ts:96]: I don't think so: In the future, it might very well be that there are authorization methods that don't return a userId here. If we still rely on this then, it's application issue.
This method is primarily useful for filling the "first argument userId: string" until we remove it.

Comment (akosyakov) [reply components/server/src/util/request-context.ts:96]: ok

File: components/server/src/api/auth-provider-service-api.ts
Comment (geropl) [reply components/server/src/api/auth-provider-service-api.ts:45]: Hm, what I did was just "replace context.user.id with ctxUserId()".

But I think you are right: ctxUserId is correct for authroization (first argument), but ownerId should be used in the shape - at least in the path where we validate it.

@AlexTugarev Could you review that section as well? Would be ideal if the two of us could sync, I could be missing sth here.

File: components/server/src/api/auth-provider-service-api.ts
Comment (AlexTugarev) [inline components/server/src/api/auth-provider-service-api.ts:63]: indeed, this can be `ownerId`.

Comment (AlexTugarev) [reply components/server/src/api/auth-provider-service-api.ts:63]: actually, let me think. I believe this needs to be the `context.user.id`/`ctxUserId` but validated to be same as `ownerId`.

here were talking about (obsolete) user-level providers, which can only be created for the current user.

Comment (geropl) [reply components/server/src/api/auth-provider-service-api.ts:63]: Yeeeeah, was thinking in the same direction.

Comment (AlexTugarev) [reply components/server/src/api/auth-provider-service-api.ts:63]: is `ctxUserId` still derived from `sub`(ject) of the JWT cookie? if so, this diff look good to me. at least, I'm not keen to change to `ownerId`.

```
oneOf owner {
  ownerId string
  orgId string
}
```
feels to be not 100% correctly defined for `createAuthProvider` params. see, there is only a way to create this type of resource with owner being the current user or the specified org.

Comment (akosyakov) [reply components/server/src/api/auth-provider-service-api.ts:63]: We can reconsider and remove oneof, just organization_id if not provided then a user?

Comment (geropl) [reply components/server/src/api/auth-provider-service-api.ts:63]: We synced on this, and decided to phase out the ownerId. This PR removes the validation/handling of it, @AlexTugarev will follow-up with a change to the proto file.

File: components/server/src/api/auth-provider-service-api.ts
Comment (akosyakov) [inline components/server/src/api/auth-provider-service-api.ts:49]: organizationId is invalid

File: components/server/src/api/server.ts
Comment (geropl) [reply components/server/src/api/server.ts:264]: There seems to be an issue indeed. Trying to debug it...

File: components/server/src/prebuilds/github-enterprise-app.ts
Comment (mustard-mh) [inline components/server/src/prebuilds/github-enterprise-app.ts:264]: How can we know that we need to do such thing? i.e. Verify if we have anywhere not add it.

Find where we use with SYSTEM_USER requestId userId?

File: components/server/src/websocket/websocket-connection-manager.ts
Comment (mustard-mh) [inline components/server/src/websocket/websocket-connection-manager.ts:387]: Can we make requestKind specify in interface? In case we do it duplicate with different format like `json-rpc`

File: components/server/src/api/auth-provider-service-api.ts
Comment (geropl) [reply components/server/src/api/auth-provider-service-api.ts:65]: The intention is to make it very easy to remove the calls in a later step again, whereever possible. To make that easier I'm accepting the minimal overhead here. :+1:

File: components/server/src/prebuilds/github-enterprise-app.ts
Comment (geropl) [reply components/server/src/prebuilds/github-enterprise-app.ts:264]: > How can we know that we need to do such thing? i.e. Verify if we have anywhere not add it.

Excellent question! I have a follow-up PR that checks this in `Authorizer`, increments a metric and logs an error. That's what we need to drive to 0 before we can use this for rollout authorization. :100:

File: components/server/src/websocket/websocket-connection-manager.ts
Comment (geropl) [reply components/server/src/websocket/websocket-connection-manager.ts:387]: Hm, ideally we should only have a single call per kind. What about this: Let's see if we can do without it, and otherwise add later?

File: components/server/src/api/server.ts
Comment (geropl) [reply components/server/src/api/server.ts]: So... it took me 2h straight to debug, fix and test in the preview. It crashed multiple times, somehow, and in the end I could not access it anymore, at all. :see_no_evil: 
Just triggered a re-deploy, should work then! :crossed_fingers:

File: components/dashboard/src/service/service.tsx
Comment (akosyakov) [inline components/dashboard/src/service/service.tsx:156]: it is enabled in prod, so can be noisy, maybe use another level? I was using Chrome debugger to verify it.

File: components/server/src/api/server.ts
Comment (geropl) [inline components/server/src/api/server.ts:263]: @akosyakov because I don't know how the resulting state machine looks like, I'm not entirely sure if we need to wrap this in `runWithSubjectId`, but I think we do.

**Comment by akosyakov** [comment]
@geropl I've enabled debugging logging `gpctl debug log server` but logs look different I would expect to see requestId and so on but getting
`{"component":"server","severity":"DEBUG","time":"2023-11-15T15:08:00.065Z","message":"public api: done","payload":{"grpc_code":"Canceled"}}`

**Comment by akosyakov** [comment]
Interesting I never see `[redis] Succesfully handled update` (debug level), but maybe i do something wrong.

**Comment by geropl** [comment]
Yeah, just found a bug in the redis handler. Pushing a fix... :roll_eyes:

**Comment by geropl** [comment]
> I've enabled debugging logging gpctl debug log server but logs look different I would expect to see requestId and so on but getting

Uh, I think `log-context.ts` is not imported anywhere anymore :laughing:

**Comment by geropl** [comment]
@akosyakov The fixes for the above are pushed and deployed.

File: components/server/src/prebuilds/github-app.ts
Comment (akosyakov) [inline components/server/src/prebuilds/github-app.ts:120]: not sure how to test it, but code wise it looks ok

File: components/server/src/prebuilds/github-app.ts
Comment (akosyakov) [inline components/server/src/prebuilds/github-app.ts:124]: in theory one can bind `req.on('aborted', () => abortController.abort())` but probably here it is not desirable, we still want to process even if client cancelled?

File: components/server/src/server.ts
Comment (akosyakov) [inline components/server/src/server.ts:151]: not sure whether we want to connect it with req.on('aborted', ...)

File: components/server/src/prebuilds/github-app.ts
Comment (geropl) [reply components/server/src/prebuilds/github-app.ts:120]: Tested that we still receive events: Preview environments are configured to use the GitHub App for that:
image

File: components/server/src/websocket/websocket-connection-manager.ts
Comment (akosyakov) [inline components/server/src/websocket/websocket-connection-manager.ts:396]: should not we await here? otherwise span in finally does not make sense?

File: components/server/src/prebuilds/github-app.ts
Comment (geropl) [reply components/server/src/prebuilds/github-app.ts:124]: Yeah. I removed that line, just to avoid any unwanted side-effects of this PR :see_no_evil:

File: components/server/src/api/server.ts
Comment (akosyakov) [reply components/server/src/api/server.ts:263]: that's fine I think :D

File: components/server/src/websocket/websocket-connection-manager.ts
Comment (geropl) [reply components/server/src/websocket/websocket-connection-manager.ts:396]: Unrelated, but totally correct! Will add.

File: components/server/src/api/server.ts
Comment (akosyakov) [inline components/server/src/api/server.ts:273]: it will call internal unexpected errors, I'm not sure whether we want to know subject id at such moment if so then it is better to wrap as well

Comment (akosyakov) [reply components/server/src/api/server.ts:273]: the same in unary cal

File: components/server/src/analytics.ts
Comment (akosyakov) [inline components/server/src/analytics.ts:137]: I wonder why don't we implement identity and track? I see segment based implementing them, should not we at least delegate?

Comment (geropl) [reply components/server/src/analytics.ts:137]: Whoops - should delegate of course :confounded:

File: components/server/src/server.ts
Comment (geropl) [reply components/server/src/server.ts:151]: :shrug: I guess it's the same as [here](https://github.com/gitpod-io/gitpod/pull/19023#discussion_r1395509942): We could argue we can safely add it because we don't check it anywhere, yet... :roll_eyes:

Comment (akosyakov) [reply components/server/src/server.ts:151]: yeah, let's not do it for now, but later we should reconsider i.e. in regards to timeouts

**Review by akosyakov — APPROVED** [review]
I could not test analytics, the rest looks good.

Please test analytics if it is alright for all kind of calls, please unhold.

/hold

File: components/server/src/api/server.ts
Comment (geropl) [reply components/server/src/api/server.ts]: Done.

File: components/server/src/websocket/websocket-connection-manager.ts
Comment (geropl) [reply components/server/src/websocket/websocket-connection-manager.ts]: Done.

File: components/server/src/analytics.ts
Comment (geropl) [reply components/server/src/analytics.ts]: Done.

File: components/dashboard/src/service/service.tsx
Comment (geropl) [reply components/dashboard/src/service/service.tsx]: Done

**Comment by geropl** [comment]
Tested analytics and it works. :heavy_check_mark: 


/unblock

File: components/server/src/api/auth-provider-service-api.ts
Comment (AlexTugarev) [inline components/server/src/api/auth-provider-service-api.ts:122]: @geropl, just realized, this actually throws if unauthenticated.

Need to address it, as it breaks Login page with gRPC enabled.

Comment (AlexTugarev) [reply components/server/src/api/auth-provider-service-api.ts:122]: https://github.com/gitpod-io/gitpod/pull/19094

Comment (geropl) [reply components/server/src/api/auth-provider-service-api.ts:122]: @AlexTugarev Yes, this one throws. If we need one that does not throw, we should name it `ctxTryUserId`, following the pattern in request-context.ts :+1:

**Pull request merged**
